core/src/apps/cardano/helpers/address_credential_policy.py

```diff
@@ -0,0 +1,196 @@
+from trezor.enums import CardanoAddressType
+
+from ..address import get_public_key_hash
+from .paths import CHAIN_STAKING_KEY, SCHEMA_PAYMENT, SCHEMA_STAKING
+from .utils import to_account_path
+
+if False:
+    from trezor.messages import CardanoAddressParametersType
+    from ..seed import Keychain
+
+
+CREDENTIAL_POLICY_SHOW = 0
```

Comment:
Thanks for suggestions, we have looked at the credential policies, dropped the unused ones and implemented something similar to what you suggest, see https://github.com/trezor/trezor-firmware/pull/1772/commits/b2a43bcf7c92d2a958a65825af44c6e2861f1be6.

Just a few notes:
1. I think that `WARN_SCRIPT = 1 | FLAG_WARNING`, `WARN_POINTER = 2 | FLAG_WARNING`, `WARN_REWARD = 3 | FLAG_WARNING` that you suggest wouldn't work, because for `p = WARN_REWARD`, checking if we should show a script warning, ie. (`p & WARN_SCRIPT == WARN_SCRIPT`) would yield falsely positive result. Therefore we need non-overlapping bits for them, something like `WARN_SCRIPT = 1 << 1 | FLAG_WARNING`, `WARN_POINTER = 1 << 2 | FLAG_WARNING`, `WARN_REWARD = 1 << 3 | FLAG_WARNING`.
2. We added only the `WARN` flag, because it is used with all other warnings. `WARN_UNUSUAL_PATH` is a normal policy just as `WARN_MISMATCH` (actually, they're somewhat symmetrical). The two can be combined and correctly checked against using `CredentialPolicy.check(...)`.
3. We have changed the actual policy generation a bit (taking into account unusual path with `REWARD` addresses and returning `WARN_MISMATCH` only when paths differ).

I hope the code should be fine now.

Reply:
> 1. I think that `WARN_SCRIPT = 1 | FLAG_WARNING`, `WARN_POINTER = 2 | FLAG_WARNING`, `WARN_REWARD = 3 | FLAG_WARNING` that you suggest wouldn't work, because for `p = WARN_REWARD`, checking if we should show a script warning

That's why I proposed to set `FLAG_WARNING = 0b1000_0000` (so `0x80`) which is a high enough bit that it won't collide with the small numbers. Your way works too, but it's slightly too atypical -- usually you'll set the flags in the high bits instead of worrying about shifting the value numbers.

Reply:
But `WARN_SCRIPT` and `WARN_REWARD` are the two colliding in my example - `WARN_SCRIPT` bits are a proper subset of `WARN_REWARD` bits.

Reply:
Do you ever need to check if `WARN_SCRIPT` and `WARN_REWARD` are true in the same credential policy?

In general, ISTM that your original implementation, as well as the updated one, is combining two concepts:

1. An **enum** of possible policies, where the result of `get_credential_policy()` is exactly one of the possible results
   The use of `WARN` as a bitflag is simply a convenience here - a way to indicate that values from 0-127 are OK statuses and values 128-255 are WARN statuses.
   I'd expect to check which one of those it is by `policy == SHOW`, `policy == WARN_SCRIPT`, etc.

2. Some _bit flags_, which indicate _additional things_ that are true about the result.
   The only one of those additional things is "is the path unusual? y/n".
   This breaks the (1) usecase a little, instead of `policy == SHOW`, you'd need to write something like `policy & POLICY_MASK == SHOW`, where `POLICY_MASK = 0xFF` and `FLAG_UNUSUAL_PATH = 0x100`.

But all together, I don't see the need for each value to be a separate flag. That would only be useful if you ever want `SHOW`, `WARN_SCRIPT` and `WARN_REWARD` to be valid at the same time.

Reply:
We've reconsidered the actual policies and `SHOW_XXX`, `WARN_SCRIPT` and `WARN_REWARD` turned out to be redundant. Maybe it would be helpful if we switched in this thread to the current policies.

These are all the policies that can appear as of now: 
```
NONE
WARN  # general / other
WARN_REWARD
WARN_NO_STAKING
WARN_MISMATCH
WARN_UNUSUAL_PATH
WARN_MISMATCH_AND_UNUSUAL_PATH
# all policies prefixed with WARN are considered warning policies
```
We want to be able to separately check:
1. Whether a policy is a warning policy.
2. Whether we should show the reward warning.
3. Whether we should show the no staking warning.
4. Whether we should show the mismatch warning.
5. Whether we should show the unusual path warning.
 
Using eg.
```
FLAG_WARN = 0b1000_0000
FLAG_UNUSUAL_PATH = 0b1_0000_0000
NONE = 0
WARN_REWARD = 1 | WARN
WARN_NO_STAKING = 2 | WARN
WARN_MISMATCH = 3 | WARN
```
would make it impossible to have a single comparison method (if I'm not mistaken):
- `p == POLICY_TO_CHECK` wouldn't work because `p = WARN_MISMATCH | FLAG_UNUSUAL_PATH` would return negative answer when comparing to `WARN_MISMATCH`
-  `p & POLICY_TO_CHECK` wouldn't work because `p = WARN_NO_STAKING` would return positive answer when comparing to `WARN_REWARD` (because of the `WARN` bit)
- `p & POLICY_TO_CHECK == POLICY_TO_CHECK` wouldn't work because `p = WARN_MISMATCH` would return positive answer when comparing to `WARN_REWARD`.

Furthermore, this solution wouldn't reflect the fact that `FLAG_UNUSUAL_PATH` should imply `FLAG_WARN`.

I understand your concern about combining the enum and bitflag concepts, but we both with @gabrielKerekes analyzed this and considered the current approach good enough while being reasonably simple (eg. we could have a pure enum with all possible policies and `should_show_reward_warning()`, `should_show_no_staking_warning()`, etc. methods, but that would be needlessly explicit). What do you think?

Reply:
resolved by a different approach